### PR TITLE
Inherit TeixeiraWaterSQE from FunctionQDepends

### LINKED ISSUE #18231: Inherit TeixeiraWaterSQE from FunctionQDepends
Depends on #18101 

[TeixeiraWaterSQE](http://docs.mantidproject.org/nightly/fitfunctions/TeixeiraWaterSQE.html)  is a Q-dependent model, thus it should inherit from base class FunctionQDepends

### PR DESCRIPTION
**Pull request by jmborr**
1. Refactoring of the code so that fit function now inherits from FunctionQDepends
2. Added doctest
3. Added system test

Try running the code below and verify that parameter DiffCoeff representing the self-diffusion constant of water at room temperature is reasonably independent of  the momentum transfer `Q`, which is different for each of the spectra of `irs26176_graphite002_red`. `Q` is 0.552 for the first spectrum and increases up to 1.8 corresponding to the last spectrum.

```
# Download data for system tests and make sure you include
# the path in the list of Data Search Directories"

""" Data corresponding to self-diffusion of bulk water at room temperature"""
resolution=Load("irs26173_graphite_res.nxs")
data=Load("irs26176_graphite002_red")

"""
We need to fix Tau parameter in 'function' because otherwise we incur in overfitting,
we choose a reasonable value for water at room temperature (Tau=1.16ps).

We  do a sequential fit to obtain DiffCoeff for each spectrum. The resulting
values are very similar to each other, indicative that the model describes the
diffusive dynamics of water at room temperature.
"""

function="""
(composite=Convolution,FixResolution=false,NumDeriv=true;
  name=TabulatedFunction,Workspace=resolution,WorkspaceIndex=0,
         Scaling=1,Shift=0,XScaling=1,ties=(Scaling=1,XScaling=1);
  (  name=DeltaFunction,Centre=0,ties=(Centre=0);
     name=TeixeiraWaterSQE,Tau=1.16,Centre=0,ties=(Centre=0,Tau=1.156)
  )
);
name=LinearBackground"""

PlotPeakByLogValue(Input="data,v0:10", Function=function, FitType="Sequential",
    StartX=-0.4, EndX=0.4, CreateOutput=True, OutputWorkspace="fit", MaxIterations=100)

"""
Look in workspace fit_Parameters, column named f0.f1.f1.DiffCoeff[Y]. Values
should fluctuate in the [1.5, 1.7] range.
"""
```

Fixes #18231.

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by stuartcampbell** [comment]
All looks good to me.. :shipit:

**Review by stuartcampbell — APPROVED** [review]
All looks good to me :shipit:

File: Framework/CurveFitting/src/Functions/TeixeiraWaterSQE.cpp [hunk 1]
```diff
@@ -88,6 +70,57 @@ void TeixeiraWaterSQE::function1D(double *out, const double *xValues,
...
+        this->function1D(&derivative[0], xValues, nData);
+      } else {
+        // numerical derivative
+        double delta =
+            cutoff[pName] > fabs(pVal * deltaF) ? cutoff[pName] : pVal * deltaF;
+        this->setActiveParameter(iP, pVal + delta);
+        this->applyTies();
+        this->function1D(&derivative[0], xValues, nData);
+        for (size_t i = 0; i < nData; i++) {
+          derivative.at(i) = (derivative.at(i) - out.at(i)) / delta;
```
Comment (quantumsteve) [inline Framework/CurveFitting/src/Functions/TeixeiraWaterSQE.cpp]: @Why are you use `std::vector::at(...)`?

File: Framework/CurveFitting/src/Functions/TeixeiraWaterSQE.cpp [hunk 2]
```diff
@@ -88,6 +70,57 @@ void TeixeiraWaterSQE::function1D(double *out, const double *xValues,
...
+        this->setActiveParameter(iP, 1.0);
+        this->applyTies();
+        this->function1D(&derivative[0], xValues, nData);
+      } else {
+        // numerical derivative
+        double delta =
+            cutoff[pName] > fabs(pVal * deltaF) ? cutoff[pName] : pVal * deltaF;
+        this->setActiveParameter(iP, pVal + delta);
+        this->applyTies();
+        this->function1D(&derivative[0], xValues, nData);
```
Comment (quantumsteve) [inline Framework/CurveFitting/src/Functions/TeixeiraWaterSQE.cpp]: use `derivative.data()`

File: Framework/CurveFitting/src/Functions/TeixeiraWaterSQE.cpp [hunk 3]
```diff
@@ -88,6 +70,57 @@ void TeixeiraWaterSQE::function1D(double *out, const double *xValues,
...
+  for (size_t iP = 0; iP < nParam; iP++) {
+    std::vector<double> derivative(nData);
+    if (this->isActive(iP)) {
+      const double pVal = this->activeParameter(iP);
+      const std::string pName = this->parameterName(iP);
+      if (pName == "Height") {
+        // exact derivative
+        this->setActiveParameter(iP, 1.0);
+        this->applyTies();
+        this->function1D(&derivative[0], xValues, nData);
```
Comment (quantumsteve) [inline Framework/CurveFitting/src/Functions/TeixeiraWaterSQE.cpp]: use `derivative.data()`

File: Framework/CurveFitting/src/Functions/TeixeiraWaterSQE.cpp [hunk 4]
```diff
@@ -88,6 +70,57 @@ void TeixeiraWaterSQE::function1D(double *out, const double *xValues,
...
+        for (size_t i = 0; i < nData; i++) {
+          derivative.at(i) = (derivative.at(i) - out.at(i)) / delta;
+        }
+      }
+      this->setActiveParameter(iP, pVal); // restore the value of the parameter
+      // fill the jacobian for this parameter
+      for (size_t i = 0; i < nData; i++) {
+        jacobian->set(i, iP, derivative.at(i));
```
Comment (quantumsteve) [inline Framework/CurveFitting/src/Functions/TeixeiraWaterSQE.cpp]: Why are you use `std::vector::at(...)`?

**Pull request merged**